### PR TITLE
Implement the device search and pairing UI

### PR DESCRIPTION
**Pull request by ekigamba**
Fixes [#fhicore/1076](https://github.com/opensrp/fhircore/issues/1076)

### CODE REVIEW

**Review by pld — APPROVED** [review]

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+        showScanningDialog()
+
+    }
+
+    fun renameWifiDirectName() {
+        val deviceName = getDeviceName(this)
+
+        // Copy the text to the Clipboard
+        val clipboard: ClipboardManager = getSystemService(CLIPBOARD_SERVICE) as ClipboardManager
+        val clip = ClipData.newPlainText("Wifi-direct name", deviceName)
```
Comment (Rkareko) [inline p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:104]: Move string to `strings.xml`

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+        clipboard.setPrimaryClip(clip)
+
+        Toast.makeText(this,  "Go to Wifi-Direct settings > Rename device and paste into the text box", Toast.LENGTH_LONG)
```
Comment (Rkareko) [inline p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:107]: Move string to `strings.xml`

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+            }
+        }
+    }
+
+    override fun handleWifiP2pDisabled() {
+        val message = "Wifi P2P: Disabled"
+        Snackbar.make(rootView, message, Snackbar.LENGTH_LONG).show()
+        /*findViewById<TextView>(R.id.wifi_p2p_enabled_value).apply {
+            text = getString(R.string.wifi_p2p_disabled_value)
+        }*/
```
Comment (Rkareko) [inline p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:257]: Remove commented out code

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+        Timber.d(message)
+    }
+
+    override fun handleWifiP2pEnabled() {
+        val message = "Wifi P2P: Enabled"
+        /*findViewById<TextView>(R.id.wifi_p2p_enabled_value).apply {
+            text = getString(R.string.wifi_p2p_enabled_value)
+        }*/
```
Comment (Rkareko) [inline p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:265]: Remove commented out code

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+        Timber.d(message)
+    }
+
+    override fun handleUnexpectedWifiP2pState(wifiState: Int) {
+        val message = "Wifi P2P: Unexpected state: $wifiState"
+        Snackbar.make(rootView, message, Snackbar.LENGTH_LONG).show()
+        /*findViewById<TextView>(R.id.wifi_p2p_enabled_value).apply {
+            text = getString(R.string.wifi_p2p_unexpected_state_value, wifiState)
+        }*/
```
Comment (Rkareko) [inline p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:274]: Remove commented out code

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+        val reason = getWifiP2pReason(reasonInt)
+        val message = "Wifi P2P: Peer discovery failed: $reason"
+        Snackbar.make(rootView, message, Snackbar.LENGTH_LONG).show()
+        Timber.d(message)
+    }
+
+    override fun handleP2pDiscoveryStarted() {
+        /*findViewById<TextView>(R.id.wifi_p2p_discovery_value).apply {
+            text = getString(R.string.wifi_p2p_discovery_started_value)
+        }*/
```
Comment (Rkareko) [inline p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:323]: Remove commented out code

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+        Timber.d("Wifi P2P: Peer discovery started")
+    }
+
+    override fun handleP2pDiscoveryStopped() {
+        /*findViewById<TextView>(R.id.wifi_p2p_discovery_value).apply {
+            text = getString(R.string.wifi_p2p_discovery_started_value)
+        }*/
```
Comment (Rkareko) [inline p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:330]: Remove commented out code

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+        Timber.d("Wifi P2P: Peer discovery stopped")
+    }
+
+    override fun handleUnexpectedWifiP2pDiscoveryState(discoveryState: Int) {
+        /*findViewById<TextView>(R.id.wifi_p2p_discovery_value).apply {
+            text = getString(R.string.wifi_p2p_unexpected_state_value, discoveryState)
+        }*/
```
Comment (Rkareko) [inline p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:337]: Remove commented out code

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+                    handleDeviceConnectionFailure(device, reason)
+                }
+            })
+        }
+    }
+
+    private fun handleDeviceConnectionSuccess(device: WifiP2pDevice) {
+        /*findViewById<TextView>(R.id.wifi_p2p_connection_value).apply {
+            text = device.deviceName
+        }*/
```
Comment (Rkareko) [inline p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:404]: Remove commented out code

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt [hunk 10]
```diff
@@ -0,0 +1,455 @@
...
+        Timber.d("Wifi P2P: Successfully connected to device: ${device.deviceAddress}")
+    }
+
+    private fun handleDeviceConnectionFailure(device: WifiP2pDevice, reasonInt: Int) {
+        val reason = getWifiP2pReason(reasonInt)
+        /*findViewById<TextView>(R.id.wifi_p2p_connection_value).apply {
+            text = getString(R.string.wifi_p2p_connection_failed_value, reason)
+        }*/
```
Comment (Rkareko) [inline p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:412]: Remove commented out code

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt [hunk 11]
```diff
@@ -0,0 +1,455 @@
...
+        Timber.d(message)
+        /*findViewById<TextView>(R.id.wifi_p2p_group_formed_value).apply {
+            text = resources.getString(if (info.groupFormed) R.string.wifi_p2p_value_yes else R.string.wifi_p2p_value_no)
+        }
+        findViewById<TextView>(R.id.wifi_p2p_group_owner_value).apply {
+            text = resources.getString(if (info.isGroupOwner) R.string.wifi_p2p_value_yes else R.string.wifi_p2p_value_no)
+        }
+        findViewById<TextView>(R.id.wifi_p2p_group_owner_address_value).apply {
+            text = if (info.groupOwnerAddress == null) resources.getString(R.string.wifi_p2p_group_owner_value_na) else info.groupOwnerAddress.hostAddress
+        }*/
```
Comment (Rkareko) [inline p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:437]: Remove commented out code

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt (hunk 1 above)
Comment (ekigamba) [reply p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:104]: Done

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt (hunk 2 above)
Comment (ekigamba) [reply p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:107]: Done

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt (hunk 3 above)
Comment (ekigamba) [reply p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:257]: Removed

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt (hunk 4 above)
Comment (ekigamba) [reply p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:265]: Removed

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt (hunk 5 above)
Comment (ekigamba) [reply p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:274]: Removed

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt (hunk 6 above)
Comment (ekigamba) [reply p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:323]: Removed

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt (hunk 7 above)
Comment (ekigamba) [reply p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:330]: Removed

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt (hunk 8 above)
Comment (ekigamba) [reply p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:337]: Removed

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt (hunk 9 above)
Comment (ekigamba) [reply p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:404]: Removed

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt (hunk 10 above)
Comment (ekigamba) [reply p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:412]: Removed

File: p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt (hunk 11 above)
Comment (ekigamba) [reply p2p-lib/src/main/java/org/smartregister/p2p/search/ui/P2PDeviceSearchActivity.kt:437]: Removed

**Review by Rkareko — APPROVED** [review]

**Pull request merged**